src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -769,26 +814,57 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                                 assert(targetReg != embMaskOp2Reg);
                                 assert(HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrinEmbMask.id));
 
-                                GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, embMaskOp1Reg);
-
-                                emitInsHelper(targetReg, maskReg, embMaskOp2Reg);
+                                if (hasShift)
+                                {
+                                    HWIntrinsicImmOpHelper helper(this, intrinEmbMask.op2, op2->AsHWIntrinsic(), 2);
+                                    for (helper.EmitBegin(); !helper.Done(); helper.EmitCaseEnd())
+                                    {
+                                        GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg,
+                                                                  embMaskOp1Reg);
+                                        GetEmitter()->emitInsSve_R_R_I(insEmbMask, emitSize, targetReg, maskReg,
+                                                                       helper.ImmValue(), embOpt, sopt);
+                                    }
+                                }
+                                else
+                                {
+                                    GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, embMaskOp1Reg);
+                                    GetEmitter()->emitIns_R_R_R(insEmbMask, emitSize, targetReg, maskReg, embMaskOp2Reg,
+                                                                embOpt, sopt);
+                                }
                             }
 
                             GetEmitter()->emitIns_R_R_R_R(INS_sve_sel, emitSize, targetReg, maskReg, targetReg,
                                                           falseReg, opt);
-                            break;
                         }
                         else if (targetReg != embMaskOp1Reg)
                         {
                             // embMaskOp1Reg is same as `falseReg`, but not same as `targetReg`. Move the
                             // `embMaskOp1Reg` i.e. `falseReg` in `targetReg`, using "unpredicated movprfx", so the
                             // subsequent `insEmbMask` operation can be merged on top of it.
-                            GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, falseReg);
-                        }
 
-                        // Finally, perform the actual "predicated" operation so that `targetReg` is the first operand
-                        // and `embMaskOp2Reg` is the second operand.
-                        emitInsHelper(targetReg, maskReg, embMaskOp2Reg);
+                            if (hasShift)
```

Comment:
can we move this duplicate code in a lambda or method?

Reply:
Done. Managed to remove three of them via use of a single `emitInsMovPrfxHelper()`